File: src/workflows/pr-lifecycle.ts
import { log, sleep, workflowInfo, ApplicationFailure } from '@temporalio/workflow';
import { advisor, cheap, heavy, heavyCodex, ciWait } from './proxies';
import type {
  AdvisorVerdict,
  CheckConflictOutput,
  CIResult,
  PRInfo,
} from '../activities';

export interface RobustPRMergeInput {
  repoFullName: string;
  workdir: string;
  branch: string;
  baseBranch: string;
  prTitle: string;
  prBody: string;
  /** Hard cap on CI-fail / conflict iterations to prevent infinite loops. */
  maxFixIterations?: number;
  /** When false, run CI / self-heal but stop before merging. Defaults to true. */
  autoMerge?: boolean;
  /**
   * Hard cap on advisor (top-model) consultations within this workflow run.
   * Each consult costs more tokens than a regular codex call, so we keep this
   * tight. Default 2: one for "should I keep self-healing?" and one reserve.
   */
  maxAdvisorConsults?: number;
  /**
   * Override for the post-merge poll interval (ms). Production uses the
   * 10-second default; tests inject a smaller value to keep wall time low.
   */
  postMergePollIntervalMs?: number;
  /** Override for the post-merge poll attempt count. Tests can lower this. */
  postMergePollAttempts?: number;
}

export interface RobustPRMergeOutput {
  prNumber: number;
  prUrl: string;
  iterations: number;
  /**
   * True only when the workflow has *observed* the PR enter the MERGED state.
   * `gh pr merge --auto` queues the merge — the workflow polls afterwards to
   * convert "merge requested" into "merge confirmed".
   */
  merged: boolean;
  /**
   * Distinguishes the terminal state for the operator:
   *  - `merged`: actually landed (mergedAt observed).
   *  - `merge-queued`: gh accepted `--auto`; merge has not yet landed within
   *    the post-merge poll window (e.g. waiting on required-up-to-date).
   *  - `auto-merge-disabled`: caller asked to stop before merging.
   *  - `closed-externally` / `merged-externally`: the PR was closed or merged
   *    by something other than this workflow during the CI loop.
   */
  outcome:
    | 'merged'
    | 'merge-queued'
    | 'auto-merge-disabled'
    | 'closed-externally'
    | 'merged-externally';
  /** Number of advisor consults actually performed. */
  advisorConsults: number;
}

const POST_MERGE_POLL_ATTEMPTS = 6;
const POST_MERGE_POLL_INTERVAL_MS = 10_000;
const CI_POLL_INTERVAL_SECONDS = 30;
const CI_MAX_WAIT_SECONDS = 60 * 60;

/**
 * Common PR lifecycle, robust against external interference (merges from
 * other PRs, human cancellation, codex producing no diff):
 *   1. Push branch + open PR.
 *   2. CI loop — on failure: pull failed logs, codex fix, push, retry.
 *      Optionally consult an Advisor before iter ≥ 2 to bail on structural
 *      failures rather than burning the iteration budget.
 *   3. Conflict loop — on conflict: codex resolve, push, return to CI loop.
 *   4. Merge: request via `gh pr merge --auto`, then poll until the PR is
 *      observed in MERGED state (or report `merge-queued` if it doesn't land
 *      promptly — typical when branch protection requires up-to-date).
 *   5. PR observed CLOSED / MERGED externally (e.g. another PR merged first
 *      and the human resolved by closing this one) — short-circuit out of
 *      the loop without throwing.
 */
export async function robustPRMergeWorkflow(
  input: RobustPRMergeInput,
): Promise<RobustPRMergeOutput> {
  const maxIters = input.maxFixIterations ?? 8;
  const autoMerge = input.autoMerge ?? true;
  const advisorBudget = new AdvisorBudget(input.maxAdvisorConsults ?? 2);
  const info = workflowInfo();

  await heavy.pushBranchActivity({
    workdir: input.workdir,
    branch: input.branch,
    setUpstream: true,
  });

  const pr: PRInfo = await cheap.createPRActivity({
    repoFullName: input.repoFullName,
    workdir: input.workdir,
    branch: input.branch,
    baseBranch: input.baseBranch,
    title: input.prTitle,
    body: input.prBody,
    draft: false,
  });
  log.info('Opened PR', { pr: pr.url, workflowId: info.workflowId });

  let iter = 0;
  while (iter < maxIters) {
    const ci = await waitForCI(pr.number, input.repoFullName);

    const externalExit = handleExternalExit(ci, pr, advisorBudget);
    if (externalExit) return externalExit;

    if (ci.status === 'failure') {
      iter += 1;
      const advice = await maybeConsultBeforeSelfHeal({
        iter,
        ci,
        input,
        advisorBudget,
      });
      if (advice === 'abort') {
        throw ApplicationFailure.create({
          message: `Advisor recommended abort after CI iter ${iter} on PR #${pr.number}`,
          type: 'AdvisorAbort',
        });
      }

      await runCISelfHeal({ iter, ci, input, advisorBudget });
      continue;
    }

    const conflict = await cheap.checkConflictActivity({
      workdir: input.workdir,
      baseBranch: input.baseBranch,
    });

    if (conflict.hasConflict) {
      iter += 1;
      await runConflictResolve({
        iter,
        conflict,
        input,
        advisorBudget,
      });
      continue;
    }

    if (!autoMerge) {
      log.info('autoMerge=false; skipping merge', { pr: pr.url, workflowId: info.workflowId });
      return {
        prNumber: pr.number,
        prUrl: pr.url,
        iterations: iter,
        merged: false,
        outcome: 'auto-merge-disabled',
        advisorConsults: advisorBudget.used(),
      };
    }

    await cheap.mergePRActivity({
      repoFullName: input.repoFullName,
      prNumber: pr.number,
      mergeMethod: 'squash',
      deleteBranch: true,
    });

    const observed = await pollUntilMerged(
      input.repoFullName,
      pr.number,
      input.postMergePollAttempts ?? POST_MERGE_POLL_ATTEMPTS,
      input.postMergePollIntervalMs ?? POST_MERGE_POLL_INTERVAL_MS,
    );
    return {
      prNumber: pr.number,
      prUrl: pr.url,
      iterations: iter,
      merged: observed === 'merged',
      outcome: observed,
      advisorConsults: advisorBudget.used(),
    };
  }

  throw ApplicationFailure.create({
    message: `PR #${pr.number} exceeded max self-heal iterations (${maxIters})`,
    type: 'MaxIterationsExceeded',
  });
}

// ──────────────────────────────────────────────────────────────────────────
// Step helpers — extracted so the main loop reads as a sequence of intents.
// All helpers are pure orchestration over Activity calls.
// ──────────────────────────────────────────────────────────────────────────

async function waitForCI(prNumber: number, repoFullName: string): Promise<CIResult> {
  return ciWait.waitForCIActivity({
    repoFullName,
    prNumber,
    pollIntervalSeconds: CI_POLL_INTERVAL_SECONDS,
    maxWaitSeconds: CI_MAX_WAIT_SECONDS,
  });
}

/**
 * If CI returns one of the externally-driven terminal states (`closed`,
 * `merged`, `timeout`), translate that into a workflow outcome — either an
 * early successful return or a thrown failure. Returns undefined when the
 * workflow should keep going (success / failure cases).
 */
function handleExternalExit(
  ci: CIResult,
  pr: PRInfo,
  advisorBudget: AdvisorBudget,
): RobustPRMergeOutput | undefined {
  if (ci.status === 'timeout') {
    throw ApplicationFailure.create({
      message: `CI did not converge within timeout for PR #${pr.number}`,
      type: 'CITimeout',
    });
  }
  if (ci.status === 'closed') {
    log.info('PR was closed externally; abandoning workflow', { pr: pr.url });
    return {
      prNumber: pr.number,
      prUrl: pr.url,
      iterations: 0,
      merged: false,
      outcome: 'closed-externally',
      advisorConsults: advisorBudget.used(),
    };
  }
  if (ci.status === 'merged') {
    log.info('PR was merged externally; treating as success', { pr: pr.url });
    return {
      prNumber: pr.number,
      prUrl: pr.url,
      iterations: 0,
      merged: true,
      outcome: 'merged-externally',
      advisorConsults: advisorBudget.used(),
    };
  }
  return undefined;
}

interface SelfHealContext {
  iter: number;
  /** Always called with `ci.status === 'failure'` — typed as the full union to keep call sites simple. */
  ci: CIResult;
  input: RobustPRMergeInput;
  advisorBudget: AdvisorBudget;
}

/**
 * Optionally consults the advisor BEFORE attempting another self-heal.
 * Triggers only when:
 *   - we're entering iter ≥ 2 (the first attempt deserves to run; the second
 *     is when "is this even fixable?" becomes worth asking), AND
 *   - the advisor budget has remaining capacity.
 *
 * Returns `'continue'` to proceed with self-heal, `'abort'` to stop, or
 * `'change-strategy'` to also proceed (the workflow doesn't have a third
 * branch yet — we surface the suggested action in logs and PR body, then
 * fall through to the normal self-heal). The verdict gating is intentionally
 * simple to keep the workflow's call graph small.
 */
async function maybeConsultBeforeSelfHeal(
  ctx: SelfHealContext,
): Promise<'continue' | 'abort' | 'change-strategy'> {
  if (ctx.iter < 2) return 'continue';
  if (!ctx.advisorBudget.canConsume()) {
    log.info('advisor budget exhausted; proceeding with self-heal', { iter: ctx.iter });
    return 'continue';
  }
  ctx.advisorBudget.consume();

  const summary = [
    `Failed jobs: ${ctx.ci.failedJobNames.slice(0, 6).join(', ') || '(none reported)'}`,
    `Failed run IDs: ${ctx.ci.failedRunIds.slice(0, 6).join(', ') || '(none)'}`,
    `Self-heal iteration about to start: ${ctx.iter}/${ctx.input.maxFixIterations ?? 8}`,
  ].join('\n');

  let verdict: AdvisorVerdict;
  let rationale = '';
  let suggestedAction: string | undefined;
  try {
    const reply = await advisor.consultAdvisorActivity({
      workdir: ctx.input.workdir,
      situation: `CI is red on PR #${ctx.input.branch}; deciding whether to attempt self-heal again.`,
      summary,
      options: [
        'retry — pattern looks transient or fixable with one more codex pass',
        'abort — failure is structural; stop and surface to a human',
        'change-strategy — try a different prompt or convert PR to draft for human review',
      ],
    });
    verdict = reply.verdict;
    rationale = reply.rationale;
    suggestedAction = reply.suggestedAction;
  } catch (err) {
    log.warn('advisor consult failed; defaulting to continue', { err: String(err) });
    return 'continue';
  }
  log.info('advisor verdict before self-heal', {
    iter: ctx.iter,
    verdict,
    rationale,
    suggestedAction,
  });
  return verdict === 'abort' ? 'abort' : 'continue';
}

async function runCISelfHeal(ctx: SelfHealContext): Promise<void> {
  const failedLogs = await collectFailedLogs(ctx.input.repoFullName, ctx.ci.failedRunIds);
  log.warn('CI failed; entering self-heal cycle', {
    iter: ctx.iter,
    failedRunIds: ctx.ci.failedRunIds,
  });
  const fix = await heavyCodex.codexActivity({
    workdir: ctx.input.workdir,
    systemPrompt:
      'You are an autonomous fix-it engineer. Identify the failing test/build from ' +
      'the CI logs and apply MINIMAL changes to make CI pass. ' +
      'Do not add unrelated improvements.',
    context: 'CI logs:\n' + failedLogs.slice(0, 64 * 1024),
    prompt: 'Apply the minimal fix to the working tree, then stop.',
  });
  await commitAndPushOrEscalate({
    iter: ctx.iter,
    input: ctx.input,
    advisorBudget: ctx.advisorBudget,
    commitMessage: `fix(ci): self-heal attempt ${ctx.iter}`,
    codexMessage: fix.message,
  });
}

interface ConflictResolveContext {
  iter: number;
  conflict: CheckConflictOutput;
  input: RobustPRMergeInput;
  advisorBudget: AdvisorBudget;
}

async function runConflictResolve(ctx: ConflictResolveContext): Promise<void> {
  log.warn('Merge conflict detected; entering resolve cycle', {
    iter: ctx.iter,
    files: ctx.conflict.conflictedFiles,
  });
  const resolve = await heavyCodex.codexActivity({
    workdir: ctx.input.workdir,
    systemPrompt:
      'You resolve git merge conflicts. Preserve intent from both sides whenever possible. ' +
      'Leave NO conflict markers in the resulting files.',
    context: 'Conflict diff:\n' + (ctx.conflict.diffSummary ?? ''),
    paths: ctx.conflict.conflictedFiles,
    prompt:
      'Resolve all merge conflicts in the working tree. ' +
      `Conflicted files: ${ctx.conflict.conflictedFiles.join(', ')}.`,
  });
  await commitAndPushOrEscalate({
    iter: ctx.iter,
    input: ctx.input,
    advisorBudget: ctx.advisorBudget,
    commitMessage: `chore(merge): resolve conflicts attempt ${ctx.iter}`,
    codexMessage: resolve.message,
  });
}

interface CommitAndPushContext {
  iter: number;
  input: RobustPRMergeInput;
  advisorBudget: AdvisorBudget;
  commitMessage: string;
  codexMessage: string;
}

/**
 * Commit the codex-applied fix and push it. When codex produces no diff at
 * all we optionally consult the advisor (if budget permits) for a logged
 * second opinion, then throw `NoFixDiff` — the advisor's verdict is recorded
 * for the operator but does not change the throw, because by then no fix
 * exists to push.
 */
async function commitAndPushOrEscalate(ctx: CommitAndPushContext): Promise<void> {
  const commit = await heavy.commitAllActivity({
    workdir: ctx.input.workdir,
    message: ctx.commitMessage,
  });

  if (!commit.committed) {
    await escalateNoDiff(ctx);
    return;
  }

  await heavy.pushBranchActivity({
    workdir: ctx.input.workdir,
    branch: ctx.input.branch,
  });
}

async function escalateNoDiff(ctx: CommitAndPushContext): Promise<never> {
  if (ctx.advisorBudget.canConsume()) {
    ctx.advisorBudget.consume();
    try {
      const reply = await advisor.consultAdvisorActivity({
        workdir: ctx.input.workdir,
        situation: 'codex reported success but produced no diff during self-heal',
        summary:
          `Iter ${ctx.iter}. Codex final message (truncated):\n` +
          ctx.codexMessage.slice(0, 1024),
        options: [
          'abort — codex misunderstands the failure; stop and request human review',
          'change-strategy — close PR and re-open with a different prompt next run',
        ],
      });
      log.warn('advisor on no-diff', {
        verdict: reply.verdict,
        rationale: reply.rationale,
        suggestedAction: reply.suggestedAction,
      });
    } catch (err) {
      log.warn('advisor consult on no-diff failed', { err: String(err) });
    }
  }
  throw ApplicationFailure.create({
    message: 'Codex reported success but produced no diff; cannot self-heal',
    type: 'NoFixDiff',
    details: [ctx.codexMessage.slice(0, 4096)],
  });
}

async function collectFailedLogs(
  repoFullName: string,
  failedRunIds: readonly string[],
): Promise<string> {
  const parts: string[] = [];
  for (const runId of failedRunIds.slice(0, 3)) {
    const part = await cheap.fetchFailedRunLogsActivity({ repoFullName, runId });
    parts.push(`### Run ${runId}\n${part}`);
  }
  return parts.join('\n\n');
}

/**
 * Poll for the actual merge to land. `gh pr merge --auto` only requests the
 * merge — when branch-protection requires "up to date", the merge can sit in
 * a queue. We poll a small number of times; if the PR has not entered MERGED
 * by then we report `merge-queued` rather than lying about success.
 */
async function pollUntilMerged(
  repoFullName: string,
  prNumber: number,
  attempts: number,
  intervalMs: number,
): Promise<'merged' | 'merge-queued'> {
  for (let attempt = 0; attempt < attempts; attempt++) {
    const observed = await cheap.observePRStateActivity({ repoFullName, prNumber });
    if (observed.state === 'MERGED') {
      log.info('PR merge observed', { prNumber, mergedAt: observed.mergedAt });
      return 'merged';
    }
    if (observed.state === 'CLOSED') {
      // gh accepted --auto, but the PR is now closed without merging. Surface
      // the surprise rather than report success.
      throw ApplicationFailure.create({
        message: `PR #${prNumber} was closed without merging after merge request was issued`,
        type: 'MergeAbandoned',
      });
    }
    await sleep(intervalMs);
  }
  log.info('PR still queued after merge request; reporting merge-queued', { prNumber });
  return 'merge-queued';
}

// ──────────────────────────────────────────────────────────────────────────
// Workflow-side state — pure, deterministic, safe in workflow code.
// ──────────────────────────────────────────────────────────────────────────

/**
 * Hard cap on how many advisor (top-model) consults a single workflow run
 * may make. The advisor is intentionally rare; if every retry triggered one
 * the cost would balloon. The counter is workflow-local so it is naturally
 * deterministic across replays.
 */
class AdvisorBudget {
  private consumed = 0;
  constructor(private readonly cap: number) {}
  canConsume(): boolean {
    return this.consumed < this.cap;
  }
  consume(): void {
    this.consumed += 1;
  }
  used(): number {
    return this.consumed;
  }
}

File: internal/workflow/pr_lifecycle.go
package workflow

import (
	"fmt"
	"log/slog"

	codexact "github.com/tamara1031/temporal-repo-steward/internal/activity/codex"
	gitact "github.com/tamara1031/temporal-repo-steward/internal/activity/git"
	ghact "github.com/tamara1031/temporal-repo-steward/internal/activity/github"
	rserrors "github.com/tamara1031/temporal-repo-steward/internal/errors"
	"go.temporal.io/sdk/workflow"
)

const (
	maxFixIterations     = 8
	postMergePollAttempts = 6
)

// QueryCIProgress is the query name for CIProgress.
const QueryCIProgress = "ci_progress"

// CIProgress is the payload returned by the "ci_progress" query handler on
// RobustPRMergeWorkflow. It lets operators track which CI self-heal iteration
// is in flight and what the last observed CI outcome was.
type CIProgress struct {
	PRNumber      int    `json:"pr_number"`
	PRURL         string `json:"pr_url"`
	Iteration     int    `json:"iteration"`
	MaxIterations int    `json:"max_iterations"`
	LastOutcome   string `json:"last_outcome,omitempty"`
}

// RobustPRMergeInput is the input to RobustPRMergeWorkflow.
type RobustPRMergeInput struct {
	RepoFullName string
	WorkDir      string
	Branch       string
	BaseBranch   string
	PRTitle      string
	PRBody       string
	SessionID    string
	AutoMerge    bool
}

// RobustPRMergeResult is the result of RobustPRMergeWorkflow.
type RobustPRMergeResult struct {
	PRNumber int
	PRURL    string
	Merged   bool
	Outcome  string // "merged" | "merge-queued" | "closed-externally" | "auto-merge-disabled"
}

// RobustPRMergeWorkflow creates a PR, waits for CI, self-heals failures, and merges.
func RobustPRMergeWorkflow(ctx workflow.Context, in RobustPRMergeInput) (RobustPRMergeResult, error) {
	ciProgress := CIProgress{MaxIterations: maxFixIterations}
	if err := workflow.SetQueryHandler(ctx, QueryCIProgress, func() (CIProgress, error) {
		return ciProgress, nil
	}); err != nil {
		return RobustPRMergeResult{}, fmt.Errorf("register ci_progress query: %w", err)
	}

	var ghActs *ghact.Activities
	var gitActs *gitact.Activities
	var codexActs *codexact.Activities

	if err := workflow.ExecuteActivity(
		workflow.WithActivityOptions(ctx, heavyGitActOpts()),
		gitActs.PushBranchActivity,
		gitact.PushInput{WorkDir: in.WorkDir, Branch: in.Branch},
	).Get(ctx, nil); err != nil {
		return RobustPRMergeResult{}, fmt.Errorf("push: %w", err)
	}

	var prResult ghact.CreatePRResult
	if err := workflow.ExecuteActivity(
		workflow.WithActivityOptions(ctx, fastGHActOpts()),
		ghActs.CreatePRActivity,
		ghact.CreatePRInput{
			WorkDir:    in.WorkDir,
			Title:      in.PRTitle,
			Body:       in.PRBody,
			BaseBranch: in.BaseBranch,
			Branch:     in.Branch,
		},
	).Get(ctx, &prResult); err != nil {
		return RobustPRMergeResult{}, fmt.Errorf("create PR: %w", err)
	}

	result := RobustPRMergeResult{
		PRNumber: prResult.Number,
		PRURL:    prResult.URL,
	}

	ciProgress.PRNumber = prResult.Number
	ciProgress.PRURL = prResult.URL

	for iteration := 0; iteration < maxFixIterations; iteration++ {
		ciProgress.Iteration = iteration

		var ciResult ghact.WaitForCIResult
		if err := workflow.ExecuteActivity(
			workflow.WithActivityOptions(ctx, ciPollActOpts()),
			ghActs.WaitForCIActivity,
			ghact.WaitForCIInput{WorkDir: in.WorkDir, PRNumber: prResult.Number},
		).Get(ctx, &ciResult); err != nil {
			return result, err
		}

		ciProgress.LastOutcome = string(ciResult.Outcome)

		switch ciResult.Outcome {
		case ghact.CIOutcomeExternallyMerged:
			result.Outcome = "merged-externally"
			result.Merged = true
			return result, nil
		case ghact.CIOutcomeExternallyClosed:
			result.Outcome = "closed-externally"
			return result, nil
		case ghact.CIOutcomeSuccess:
			if !in.AutoMerge {
				result.Outcome = "auto-merge-disabled"
				return result, nil
			}

			// Before merging, check whether the PR has accumulated conflict markers
			// since CI was triggered (e.g. a concurrent push to the base branch).
			var mergeCheck ghact.CheckMergeableResult
			_ = workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, fastGHActOpts()),
				ghActs.CheckMergeableActivity,
				ghact.CheckMergeableInput{WorkDir: in.WorkDir, PRNumber: prResult.Number},
			).Get(ctx, &mergeCheck)

			if mergeCheck.Status == ghact.MergeStateDirty {
				if iteration == maxFixIterations-1 {
					return result, rserrors.NewMaxIterations()
				}
				slog.Info("PR has merge conflicts, attempting resolution", "iteration", iteration)

				var mergeResult gitact.FetchAndMergeResult
				if err := workflow.ExecuteActivity(
					workflow.WithActivityOptions(ctx, heavyGitActOpts()),
					gitActs.FetchAndStartMergeActivity,
					gitact.FetchAndMergeInput{WorkDir: in.WorkDir, BaseBranch: in.BaseBranch},
				).Get(ctx, &mergeResult); err != nil {
					continue // fetch/merge activity failed; retry on next iteration
				}

				if mergeResult.HasConflict {
					_ = workflow.ExecuteActivity(
						workflow.WithActivityOptions(ctx, longCodexActOpts()),
						codexActs.ChatActivity,
						codexact.ChatInput{
							SessionID: in.SessionID,
							Message:   "Resolve all merge conflict markers in the working directory. Apply conflict resolution only — no other changes.",
						},
					).Get(ctx, nil)

					var conflictSHA string
					if err := workflow.ExecuteActivity(
						workflow.WithActivityOptions(ctx, fastGHActOpts()),
						gitActs.CommitAllActivity,
						gitact.CommitAllInput{
							WorkDir: in.WorkDir,
							Message: fmt.Sprintf("fix: resolve merge conflicts (iteration %d)", iteration+1),
						},
					).Get(ctx, &conflictSHA); err != nil {
						// codex produced no diff; abort the in-progress merge and retry.
						_ = workflow.ExecuteActivity(
							workflow.WithActivityOptions(ctx, fastGHActOpts()),
							gitActs.AbortMergeActivity,
							in.WorkDir,
						).Get(ctx, nil)
						continue
					}
				}

				// Either the merge was clean or conflicts were resolved; push and re-queue CI.
				if err := workflow.ExecuteActivity(
					workflow.WithActivityOptions(ctx, heavyGitActOpts()),
					gitActs.PushBranchActivity,
					gitact.PushInput{WorkDir: in.WorkDir, Branch: in.Branch, Force: true},
				).Get(ctx, nil); err != nil {
					return result, fmt.Errorf("push conflict resolution: %w", err)
				}
				continue
			}

			// PR is CLEAN (or UNKNOWN — proceed optimistically).
			if err := workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, fastGHActOpts()),
				ghActs.MergePRActivity,
				ghact.MergePRInput{WorkDir: in.WorkDir, PRNumber: prResult.Number},
			).Get(ctx, nil); err != nil {
				return result, fmt.Errorf("merge: %w", err)
			}
			var finalOutcome ghact.CIOutcome
			_ = workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, fastGHActOpts()),
				ghActs.ObservePRStateActivity,
				ghact.ObservePRStateInput{WorkDir: in.WorkDir, PRNumber: prResult.Number, Attempts: postMergePollAttempts},
			).Get(ctx, &finalOutcome)
			result.Merged = true
			result.Outcome = string(finalOutcome)
			return result, nil
		case ghact.CIOutcomeFailure:
			if iteration == maxFixIterations-1 {
				return result, rserrors.NewMaxIterations()
			}
			slog.Info("CI failed, attempting self-heal", "iteration", iteration)

			var failLogs string
			_ = workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, fastGHActOpts()),
				ghActs.FetchFailedRunLogsActivity,
				ghact.FetchFailedRunLogsInput{
					WorkDir:       in.WorkDir,
					PRNumber:      prResult.Number,
					FailedRunURLs: ciResult.FailedRuns,
				},
			).Get(ctx, &failLogs)

			var fixResult codexact.ChatResult
			if err := workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, longCodexActOpts()),
				codexActs.ChatActivity,
				codexact.ChatInput{
					SessionID: in.SessionID,
					Message:   fmt.Sprintf("Fix this CI failure. Apply the fix to the files in the working directory.\n\nFailed CI logs:\n%s", failLogs),
				},
			).Get(ctx, &fixResult); err != nil {
				continue
			}

			var commitSHA string
			if err := workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, fastGHActOpts()),
				gitActs.CommitAllActivity,
				gitact.CommitAllInput{WorkDir: in.WorkDir, Message: fmt.Sprintf("fix: CI self-heal (iteration %d)", iteration+1)},
			).Get(ctx, &commitSHA); err != nil {
				return result, rserrors.NewNoFixDiff()
			}
			if err := workflow.ExecuteActivity(
				workflow.WithActivityOptions(ctx, heavyGitActOpts()),
				gitActs.PushBranchActivity,
				gitact.PushInput{WorkDir: in.WorkDir, Branch: in.Branch, Force: true},
			).Get(ctx, nil); err != nil {
				return result, fmt.Errorf("push fix: %w", err)
			}
		}
	}

	return result, rserrors.NewMaxIterations()
}
